wire-ios-data-model/Source/MLS/MLSService.swift

```diff
@@ -815,6 +817,60 @@ public final class MLSService: MLSServiceInterface {
         }
     }
 
+    // MARK: - Out-of-sync conversations
+
+    public func repairOutOfSyncConversations() {
+        guard let context = self.context else { return }
+        logger.info("repairing out of sync conversations if needed")
+
+        outOfSyncConversations(in: context).forEach { conversation in
+            Task {
+                do {
+                    try await repairConversation(conversation)
+                    logger.info("repaired out of sync conversation")
+                } catch {
+                    logger.warn("failed to repair out of sync conversation: \(String(describing: error))")
+                }
+            }
+        }
+    }
+
+    enum ConversationRepairError: Error {
+        case missingGroupID
+    }
+
+    private func repairConversation(_ conversation: ZMConversation) async throws {
+        guard let groupID = conversation.mlsGroupID else {
+            throw ConversationRepairError.missingGroupID
+        }
+
+        try await joinGroup(with: groupID)
+    }
+
+    private func outOfSyncConversations(in context: NSManagedObjectContext) -> [ZMConversation] {
+        return coreCrypto.perform { coreCrypto in
+            return ZMConversation.fetchMLSConversations(in: context).filter {
+                isConversationOutOfSync($0, coreCrypto: coreCrypto)
+            }
+        }
+    }
+
+    private func isConversationOutOfSync(
+        _ conversation: ZMConversation,
+        coreCrypto: CoreCryptoProtocol
+    ) -> Bool {
+        guard let groupID = conversation.mlsGroupID else {
+            return false
+        }
+
+        guard let epoch = try? coreCrypto.conversationEpoch(conversationId: groupID.bytes) else {
+            // TODO: Decide if failing to fetch the epoch is considered being out of sync
```

Comment:
any decision on this?

Reply:
not really, I guess it depends on why we fail to fetch the conversation epoch. It may be that rejoining the conversation is not the solution.

Reply:
according to the [documentation](https://wireapp.github.io/core-crypto/core_crypto/struct.CoreCrypto.html#method.conversation_epoch), if there's an error it's related to the conversation not being found, which doesn't mean that it's out of sync. So I updated the logic here [3886e0f](https://github.com/wireapp/wire-ios-mono/pull/429/commits/3886e0f959a9f01a7e39fab6e5e1f539b79192a2)